chalicelib/services/InsightsService.py

```diff
@@ -0,0 +1,181 @@
+# TO BE COMPLETED: create service to perform analytics (used in API...)
+from chalicelib.db import db
+
+class InsightsService:
+    def __init__(self):
+        pass
+
+    def get_insights_from_listing(self, id: str):
+        ''' driver function of insights (returns both `dashboard` and `distribution`) '''
+
+        # fetch applicants from `get_applicants` endpoint in `db.py`
+        data = db.get_applicants(table_name="zap-applications", listing_id=id)
+
+        # call helper functions
+        # NOTE: `get_dashboard_insights` updates the data object to ensure all majors/minors are Title() cased
+        dashboard = InsightsService._get_dashboard_insights(data)
+        distribution = InsightsService._get_pie_chart_insights(data)
+
+        return dashboard, distribution
+
+    # private method (kinda)
+    def _get_dashboard_insights(data):
+        # initialize metrics
+        majors = {}
+        grad_years = {}
+        num_applicants = len(data)
+        avg_gpa = 0
+        count_gpa = 0
+
+        dashboard = {
+            "applicantCount": 0,
+            "avgGpa": "N/A",
+            "commonMajor": "N/A",
+            "commonGradYear": "N/A",
+        }
+
+        if num_applicants < 1:
+            return dashboard
+
+        # iterate over each applicant and perform analytics
+        for applicant in data:
+            # convert major/minor to title case
+            applicant["major"] = applicant["major"].title()
+            applicant["minor"] = applicant["minor"].title()
+
+            gpa, grad_year, major = applicant["gpa"], applicant["gradYear"], applicant["major"]
+
+            # attempt conversions (if fail, then skip)
+            try:
+                float_gpa = float(gpa)
+                avg_gpa += float_gpa
+                count_gpa += 1
+            except ValueError:
+                print("skipping gpa: ", gpa)
+                pass
+            try:
+                float_grad = float(grad_year)
+                if float_grad in grad_years:
+                    grad_years[float_grad] += 1
+                else:
+                    grad_years[float_grad] = 1
+            except ValueError:
+                print("skipping gradYear: ", grad_year)
+                pass
+            
+            # parse majors (if non-empty)
+            if major:
+                if major in majors:
+                    majors[major] += 1
+                else:
+                    majors[major] = 1
+        
+        avg_gpa /= count_gpa
+        common_major, count_common_major = "", 0
+        common_grad_year, count_common_grad_year = "", 0
+
+        # update most common major
+        for major, freq in majors.items():
```

Comment:
some cool tips: `max(data.items(), key=lambda x: x[1])` 

you could do something like this and manipulate this data to get the maximum/most common item of a dictionary! (a simple one liner)

Reply:
dang, man's pulled out the lambda funcion 😂

Reply:
damn that actually just cut like 15 lines of code lmaoooo